### PR TITLE
Faster G2 cofactor clearing and membership checks

### PR DESCRIPTION
**Pull request by benjaminion**
## PR Description

Implements faster algorithms for G2 cofactor clearing in hash-to-G2, and faster G2 membership tests all round. The membership test uses the method in section 3.1 of https://eprint.iacr.org/2019/814.pdf

The speed of hash-to-G2 is more than doubled. Syncing the first 12k blocks of what remains of Schlesi is ~16% faster. The performance improvement will be larger on larger networks.

### CODE REVIEW

**Comment by Nashatyrev** [comment]
x1.4 block import speed up
Before:
```
Benchmark                              (validatorsCount)  Mode  Cnt  Score   Error  Units
TransitionBenchmark.Block.importBlock              32768    ss  250  0,108 ± 0,004   s/op
```
After:
```
Benchmark                              (validatorsCount)  Mode  Cnt  Score   Error  Units
TransitionBenchmark.Block.importBlock              32768    ss  250  0,078 ± 0,003   s/op
```

File: bls/src/main/java/tech/pegasys/teku/bls/hashToG2/Helper.java [hunk 1]
```diff
@@ -54,14 +58,25 @@ static boolean isOnCurve(JacobianPoint p) {
   /**
    * Tests whether the given point lies in the G2 group.
    *
-   * <p>Once the GLV patent has expired we can replace the qChain with the endomorphism here:
-   * https://eprint.iacr.org/2019/814.pdf
-   *
    * @param p a JacobianPoint
    * @return true if the point is in G2, false otherwise
    */
   static boolean isInG2(JacobianPoint p) {
```
Comment (mbaxter) [inline bls/src/main/java/tech/pegasys/teku/bls/hashToG2/Helper.java:64]: (nit) Might be a good idea to make the methods that are directly used in production `public`, and either mark the other methods `private` or keep them package-private and annotate them with `@VisibleForTesting`.  This just makes it easier to see which methods are internal-only vs used externally.

File: bls/src/test/java/tech/pegasys/teku/bls/hashToG2/HelperTest.java [hunk 2]
```diff
@@ -206,8 +212,32 @@ void clearH2Test() {
...
+                "0x05594bb289f0ebfd8fa3f020c6e1eaf4c49b97d8ccaf3470a3a02da4b3e7104778105bd6c7e0caf97206c77a8b501d4d",
+                "0x0625151f905fad40eb0e2b9b0a46d9afe531256c6d5e39897a27d94700f037a761a741d11275180bd18e620289e02a16"));
+    assertEquals(clear_h2_slow(a), clear_h2(a));
+  }
+
+  @Test
+  void G2MapToInfinityTest() {
+    // The generator point of G2 is certainly in G2
+    JacobianPoint g2Generator = new JacobianPoint(ECP2.generator());
+    assertTrue(g2_map_to_infinity(g2Generator).isInfinity());
```
Comment (mbaxter) [inline bls/src/test/java/tech/pegasys/teku/bls/hashToG2/HelperTest.java:235]: (optional) Might be worth running this a few times, although looks like we have [another test](https://github.com/PegaSysEng/teku/blob/0dc3429662f2c8d4aba1798d075154d748a3059b/bls/src/test/java/tech/pegasys/teku/bls/hashToG2/hashToCurveTest.java#L37-L43) that basically does this already.

**Review by mbaxter — APPROVED** [review]
LGTM

File: bls/src/main/java/tech/pegasys/teku/bls/hashToG2/Helper.java (hunk 1 above)
Comment (benjaminion) [reply bls/src/main/java/tech/pegasys/teku/bls/hashToG2/Helper.java:64]: I'll be doing another pass through everything for the v0.12 update, and will consider holding my nose and adding `@VisibleForTesting` to everything. Very little can be marked `private` currently as just about every method has a corresponding unit test.

This is another reason I try not to write too much Java 😛.

File: bls/src/test/java/tech/pegasys/teku/bls/hashToG2/HelperTest.java (hunk 2 above)
Comment (benjaminion) [reply bls/src/test/java/tech/pegasys/teku/bls/hashToG2/HelperTest.java:235]: Yes, this is just a quick unit test to catch if it get's broken somehow. The smoke test exercises things more thoroughly.

**Pull request merged**